hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java

```diff
@@ -328,13 +333,40 @@ private void init() throws IOException {
     writerSchema = addPartitionFields(writerSchema, partitioningFields);
     List<String> projectionFields = orderFields(jobConf.get(ColumnProjectionUtils.READ_COLUMN_NAMES_CONF_STR),
         jobConf.get(ColumnProjectionUtils.READ_COLUMN_IDS_CONF_STR), partitioningFields);
+
+    Map<String, Field> schemaFieldsMap = getNameToFieldMap(writerSchema);
+    hiveSchema = constructHiveOrderedSchema(writerSchema, schemaFieldsMap);
     // TODO(vc): In the future, the reader schema should be updated based on log files & be able
     // to null out fields not present before
-    readerSchema = generateProjectionSchema(writerSchema, projectionFields);
+
+    readerSchema = generateProjectionSchema(writerSchema, schemaFieldsMap, projectionFields);
     LOG.info(String.format("About to read compacted logs %s for base split %s, projecting cols %s",
         split.getDeltaFilePaths(), split.getPath(), projectionFields));
   }
 
+  private Schema constructHiveOrderedSchema(Schema writerSchema, Map<String, Field> schemaFieldsMap) {
+    String hiveColumnString = jobConf.get("columns");
```

Comment:
can we pull "columns" into its own string constant with doc on what value is expected here? I also had a question on how this is different from `ColumnProjectionUtils.READ_COLUMN_NAMES_CONF_STR`

Reply:
Sure I can make it a constant with some doc.

`columns` is different because it has names of all the columns in the hive table. Whereas `ColumnProjectionUtils.READ_COLUMN_NAMES_CONF_STR` has only those that are required to be read/selected as part of the `query` that is being run.

Reply:
Cool. Did not remember seeing this before.